api/src/services/application-csv-export.service.ts

```diff
@@ -815,15 +698,380 @@ export class ApplicationCsvExporterService
+                        if (!programs) {
+                          programs =
+                            app.programs as unknown as ApplicationMultiselectQuestion[];
+                        }
+                        parsePreference = false;
+                        // there aren't typically many programs, but if there, then a object map should be created and used
+                        const program = programs.find(
+                          (preference) => preference.key === curr,
+                        );
+                        multiselectQuestionValue = true;
+                        return program;
+                      }
+
+                      // sets parsePreference to true, for the next iteration
+                      if (curr === 'preferences') {
+                        parsePreference = true;
+                      } else if (curr === 'programs') {
+                        parseProgram = true;
+                      }
+
+                      if (acc === null || acc === undefined) {
+                        return '';
+                      }
+
+                      // handles working with arrays, e.g. householdMember.0.firstName
+                      if (!isNaN(Number(curr))) {
+                        const index = Number(curr);
+                        return acc[index];
+                      }
+
+                      return acc[curr];
+                    }, app);
+                    value =
+                      value === undefined ? '' : value === null ? '' : value;
+                    if (header.format) {
+                      value = header.format(value);
+                    }
+
+                    row += value
+                      ? `"${value.toString().replace(/"/g, `""`)}"`
+                      : '';
+                    if (index < csvHeaders.length - 1) {
+                      row += ',';
+                    }
+                  });
+                  row += '\n';
+                });
+                resolve(row);
+              }),
+            );
+          }
+          const resolvedArray = await Promise.all(promiseArray);
+          // now loop over batched row data and write them to file
+          resolvedArray.forEach((row) => {
+            try {
+              writableStream.write(row);
+            } catch (e) {
+              console.log('writeStream write error = ', e);
+              writableStream.once('drain', () => {
+                console.log('drain buffer');
+                writableStream.write(row + '\n');
+              });
+            }
+          });
+          writableStream.end();
+        });
+    });
+  }
+
   async lotteryRandomizer(
     listingId: string,
     applications: Application[],
     preferencesOnListing: MultiselectQuestion[],
   ): Promise<void> {
     // remove duplicates
-    let filteredApplications = applications.filter(
-      (application) => !application.markedAsDuplicate,
-    );
+    let filteredApplications = applications;
```

Comment:
This isn't the line that is failing, but when I use HBA prod data it is failing on line 1091 with the createMany because of "Unique constraint failed on the fields: (`listing_id`,`application_id`)"

Reply:
Is this because when there are more than one multiselect question on the listing there would be two records with the same listing id and application id?

Reply:
ooooo you're right, I'll adjust the schema and generate a new migration to add `multiselectQuestionId` as part of the id 

the other option is I can add a unique id column that we use as the primary id and then add indices for `listing id` and `application id` for rapid look up